### PR TITLE
[legend pt1] legend outside plot area

### PR DESCRIPTION
**Pull request by bdeket**
PR to begin solving #49 
This is a first attempt to separate legend and function calculation (2D)

The nastiest change is for contour and contour-intervals. The legend(label) and function calculation are quite interwoven for these procedures.

With this change it should be possible to get the legend entries at any point, provided the bounds for the plot area are known (necessary for contour/-intervals). 

For contour/-intervals it uses internal state to prevent calculating the function twice.

### CODE REVIEW

**Review by alex-hhh — COMMENTED** [review]
Thanks for starting the work on the #49 issue.  The approach looks good and I only had a minor suggestion about returning two values from a function instead of a list -- I think this will improve code clarity.

Also, I think it would be a good idea to split the solving of the problem in several smaller pull-requests, since they are easier to review.   So, unless you object, I would like to merge this pull request once my comments are addressed.

File: plot-lib/plot/private/plot2d/contour.rkt [hunk 1]
```diff
@@ -72,54 +67,77 @@
...
+          (set! last-zs zs)
+          (set! last-labels labels)
+          (list sample zs labels)]
+         [else
+          (set! last-zs     empty)
+          (set! last-labels empty)
+          (list sample empty empty)])]))
+
+  (list
+   (and label (λ ([rect : Rect])
```
[the remark this answers is not in the document]
Comment (bdeket) [reply plot-lib/plot/private/plot2d/contour.rkt:116]: Yes

**Comment by bdeket** [comment]
If you think it's ok I would like to implement the same change for the 3D case in this PR. So both stay in line. But either way is fine.

**Comment by alex-hhh** [comment]
This looks good to me.  Is it ready to merge?

**Comment by bdeket** [comment]
I think so. Yes.

**Comment by bdeket** [comment]
This PR is ready.

**Pull request merged**